kernel/crates/kernel_elfloader/src/lib.rs

```diff
@@ -193,3 +193,576 @@ where
+                    let allocations = self.allocations.borrow();
+                    while allocations.contains_key(&candidate) {
+                        candidate += 0x1000;
+                    }
+                    VirtAddr::new(candidate)
+                }
+            };
+
+            let mut data = vec![0u8; layout.size()];
+            // Initialize with a pattern to detect uninitialized memory
+            data.fill(0xCC);
+
+            let alloc = MockAllocation {
+                data,
+                layout,
+                location: addr,
+            };
+
+            self.allocations
+                .borrow_mut()
+                .insert(addr.as_u64(), alloc.clone());
+            Some(alloc)
+        }
+
+        fn make_executable(
+            &mut self,
+            allocation: Self::WritableAllocation,
+        ) -> Result<Self::ExecutableAllocation, Self::WritableAllocation> {
+            if self.should_fail_make_executable {
+                return Err(allocation);
+            }
+            Ok(allocation)
+        }
+
+        fn make_writable(
+            &mut self,
+            allocation: Self::ExecutableAllocation,
+        ) -> Result<Self::WritableAllocation, Self::ExecutableAllocation> {
+            Ok(allocation)
+        }
+
+        fn make_readonly(
+            &mut self,
+            allocation: Self::WritableAllocation,
+        ) -> Result<Self::ReadonlyAllocation, Self::WritableAllocation> {
+            if self.should_fail_make_readonly {
+                return Err(allocation);
+            }
+            Ok(allocation)
+        }
+    }
+
+    // Aligned buffer for ELF header data to satisfy zerocopy alignment requirements
+    #[repr(align(8))]
+    struct AlignedElfData([u8; 64]);
+
+    impl AlignedElfData {
+        fn as_slice(&self) -> &[u8] {
+            &self.0
+        }
+
+        fn as_mut_slice(&mut self) -> &mut [u8] {
+            &mut self.0
+        }
+    }
+
+    // Helper to create an aligned Vec for ELF data
+    // Uses proper memory layout to ensure 8-byte alignment required by zerocopy
+    fn create_aligned_vec(size: usize) -> Vec<u64> {
+        // Allocate as Vec<u64> to ensure 8-byte alignment, then reinterpret as bytes
+        let num_u64s = (size + 7) / 8; // Round up to nearest u64
+        let vec = vec![0u64; num_u64s];
+        vec
+    }
+
+    // Helper to get byte slice from aligned vec
+    fn aligned_vec_as_bytes(vec: &[u64], size: usize) -> &[u8] {
+        let bytes = unsafe { core::slice::from_raw_parts(vec.as_ptr() as *const u8, size) };
+        bytes
+    }
+
+    // Helper to get mutable byte slice from aligned vec
+    fn aligned_vec_as_bytes_mut(vec: &mut [u64], size: usize) -> &mut [u8] {
+        let bytes = unsafe { core::slice::from_raw_parts_mut(vec.as_mut_ptr() as *mut u8, size) };
+        bytes
+    }
+
+    // Helper to create minimal valid ELF header
+    fn create_minimal_elf_header() -> AlignedElfData {
```

Comment:
This `tests` module seems to have a lot of code duplicated in `file.rs`. please make sure that this is not the case. Maybe provide `AlignedElfData` as struct in `kernel_elfloader`, that way users of this crate can just use that struct to make sure their data is aligned.